fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biomes/PredicatedTransformer.java

```diff
@@ -26,19 +26,61 @@
 final class PredicatedTransformer {
 	private final List<PredicatedBiomeEntry> predicates = new ArrayList<>();
 
-	void addPredicatedBiome(Biome biome, BiPredicate<Biome, LayerRandomnessSource> predicate) {
-		predicates.add(new PredicatedBiomeEntry(biome, predicate));
+	void addPredicatedBiome(Biome biome, BiPredicate<Biome, LayerRandomnessSource> predicate, double chance) {
+		predicates.add(new PredicatedBiomeEntry(biome, predicate, chance));
 	}
 
 	Biome transform(Biome biome, LayerRandomnessSource random, Biome[] borders) {
-		for (PredicatedBiomeEntry predicate : predicates) {
+		if (predicates.isEmpty()) {
+			return biome;
+		} else if (predicates.size() == 1) {
+			PredicatedBiomeEntry predicate = predicates.get(0);
 			for (Biome border : borders) {
 				if (predicate.test(border, random)) {
 					return predicate.getBiome();
 				}
 			}
+			return biome;
 		}
+		
+		List<PredicatedBiomeEntry> truePredicates = new ArrayList<>();
+		double currentTotal = 0.0D;
+		
+		for (PredicatedBiomeEntry predicate : predicates) {
+			for (Biome border : borders) {
+				if (predicate.test(border, random)) {
+					truePredicates.add(predicate);
+					
+					currentTotal += predicate.getWeight();
+					predicate.setUpperWeightBound(currentTotal);
+					continue;
```

Comment:
Do you mean to continue the outer loop here? If yes, you need a label on the outer loop.

Reply:
oops... yeah lol